sound/soc/sof/intel/hda-dai.c

```diff
@@ -383,16 +417,37 @@ static int hda_dai_hw_free(struct snd_pcm_substream *substream,
 		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
 	}
 
-	snd_soc_dai_set_dma_data(dai, substream, NULL);
+	snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
 	snd_hdac_ext_stream_release(link_dev, HDAC_EXT_STREAM_TYPE_LINK);
 	link_dev->link_prepared = 0;
```

Comment:
@plbossart the change looks good but to be truly balanced in terms of splitting DAI ops, DAI link pos and DAI link DMA ops, shouldnt the above 3 lines be called from link_dma_free() ?

Reply:
@ranj063 did you mean introducing a new function called hda_link_dma_free()?

If yes, I find it quite hard to identify what is related to the dailink and what is related to DMA proper.

Reply:
hmm yes, it is a hard split but the converse of the above are in dma_params() so maybe these can bein the new function